File: src/main/java/me/anonymoussoftware/vacancymanager/model/Vacancy.java
package me.anonymoussoftware.vacancymanager.model;

import org.json.JSONObject;

import lombok.AllArgsConstructor;
import lombok.Data;

@Data
@AllArgsConstructor
public class Vacancy {

    private final int id;
    private boolean isBanned;
    private final String name;
    private final Employer employer;
    private final Area area;
    private final Snippet snippet;
    private final Salary salary;
    private final String url;
    private String description;

    public static Vacancy fromJson(JSONObject json) {
        return new Vacancy(json.getInt("id"), //
                json.optBoolean("isBanned"), //
                json.getString("name"), //
                Employer.fromJson(json.getJSONObject("employer")), //
                Area.fromJson(json.getJSONObject("area")), //
                Snippet.fromJson(json.optJSONObject("snippet")), //
                Salary.fromJson(json.optJSONObject("salary")), //
                json.optString("alternate_url"), //
                json.optString("description"));
    }

    @Override
    public String toString() {
        return this.name + " (" + this.employer.getName() + ")";
    }

    public JSONObject toJson(boolean writeEmployerBannedStatus) {
        JSONObject result = new JSONObject();
        result.put("id", this.id);
        result.put("isBanned", this.isBanned);
        result.put("name", this.name);
        if (this.employer != null) {
            result.put("employer", this.employer.toJson(writeEmployerBannedStatus));
        }
        if (this.area != null) {
            result.put("area", this.area.toJson());
        }
        if (this.snippet != null) {
            result.put("snippet", this.snippet.toJson());
        }
        if (this.salary != null) {
            result.put("salary", this.salary.toJson());
        }
        if (this.url != null && !this.url.isEmpty()) {
            result.put("alternate_url", this.url);
        }
        if (this.description != null && !this.description.isEmpty()) {
            result.put("description", this.description);
        }
        return result;
    }
}

File: src/main/kotlin/me/anonymoussoftware/vacancymanager/model/Vacancy.kt
package me.anonymoussoftware.vacancymanager.model

import org.json.JSONObject

class Vacancy(override val id: Int,
              override val name: String,
              var isBanned: Boolean,
              val employer: Employer,
              val area: Area,
              val snippet: Snippet?,
              val salary: Salary?,
              val url: String,
              var description: String?) : ModelWithNameAndId {

    override fun toString(): String {
        return this.name + " (" + this.employer.name + ")"
    }

    fun toJson(writeEmployerBannedStatus: Boolean) = JSONObject().apply {
        put("id", id)
        put("isBanned", isBanned)
        put("name", name)
        put("employer", employer.toJson(writeEmployerBannedStatus))
        put("area", area.toJson())
        if (snippet != null) {
            put("snippet", snippet.toJson())
        }
        if (salary != null) {
            put("salary", salary.toJson())
        }
        if (url.isNotEmpty()) {
            put("alternate_url", url)
        }
        if (description != null && description!!.isNotEmpty()) {
            put("description", description)
        }
    }

    companion object : JsonDeserializer<Vacancy> {

        override fun fromJson(json: JSONObject): Vacancy {
            val salary = json.optJSONObject("salary")
            val snippet = json.optJSONObject("snippet")
            return Vacancy(
                json.getInt("id"), //
                json.getString("name"), //
                json.optBoolean("isBanned"), //
                Employer.fromJson(json.getJSONObject("employer")), //
                Area.fromJson(json.getJSONObject("area")), //
                if (snippet != null) Snippet.fromJson(snippet) else null, //
                if (salary != null) Salary.fromJson(salary) else null, //
                json.optString("alternate_url"), //
                json.optString("description")
            )
        }
    }
}
